enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -351,79 +351,67 @@ impl<'a> Handler<'a> {
     pub fn readv(&mut self) -> u64 {
         self.trace("readv", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let trusted = unsafe {
-            from_raw_parts_mut(
-                self.aex.gpr.rsi.raw() as *mut libc::iovec,
-                self.aex.gpr.rdx.raw() as usize,
-            )
-        };
+        let tiovec: &mut [libc::iovec] =
+            unsafe { self.aex.gpr.rsi.as_slice_mut(self.aex.gpr.rdx.into()) };
 
-        // Add up total size of buffers and size of iovec array.
-        let bufsize = trusted.iter().fold(0, |a, e| a + e.iov_len);
-        let iovecsize = size_of::<libc::iovec>() * trusted.len();
-        let size = bufsize + iovecsize;
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size as u64) {
-            Err(errno) => return errno as u64,
-            Ok(map) => unsafe { from_raw_parts_mut(map, size as usize) },
+        // Allocate the correct number of iovecs in untrusted memory.
+        let cursor = self.block.cursor();
+        let uiovec: &mut [libc::iovec] = match unsafe { cursor.alloc(tiovec.len()) } {
+            Ok(slice) => slice,
+            Err(_) => return -libc::EMSGSIZE as u64,
         };
```

Comment:
Using `copy_slice()` here makes the loop below smaller because you don't need to set `u.iov_len`. You just patch up the `u.iov_base`.

Reply:
@npmccallum Re: this and the above comment, I tried to implement this but ran into a ball-of-yarn problem. Because of how my code is organized, I can't get rid of the explicit typing for the `iovec` array; and trying to introduces other borrowing errors. Seems to make sense to solve this problem with your PR?

Reply:
@npmccallum Actually, I think part of your PR may fix this problem for me (returning `mut` from `copy_slice()`).